### DISCUSSION TITLE
Save Chart

### DISCUSSION THREAD

**Discussion by cotton1234** [BennyThadikaran/stock-pattern#8] [Ideas]
Hello Benny,

I have a suggestion when you are saving the list of stock in bull.json file and then we need to run the same to see the image.

If you can save the image also it will be helpful, as we need to revisit the stock later and reference the same. We can store all the images in the current date folder.

Thanks
Saurabh

**Comment by BennyThadikaran** [comment]
You dont need to re-run the scan to see the charts. 

Once the json file is generated, run **`py init.py --plot bull.json`** to plot the charts. You can use the **`n`** and **`p`** key on the chart to go forward and backward in the charts.

If you want to jump to a particular count in the result, **`py init.py --plot bull.json --idx 10`**

I am also considering adding a feature where you jump to a count from within the chart. Like type a number and press `j` to jump to that chart. Alternatively press j to pull up an index of all symbols which you use to switch charts.

If you're have particular interest in a chart, the plot window has a save button.

A save chart option can also be implemented `py init.py --save`. I wont make it the default option since generating images is a bit cpu intensive and not everyone may desire it.

I have yet to update the usage docs on the README so just giving you all the options. Implementing this is very simple.

**Reply by cotton1234** [reply]
Fine Benny. If you can implement to save all the chart, that would be great as I keep a reference of the chart when i make a position in stock. So if it fails, i need to check where I went wrong.  

Thanks
Saurabh

**Comment by BennyThadikaran** [comment]
I was thinking about the implementation, here are my thoughts

1. Add a config entry `save_folder` with the folder path in `user.json`. That way you have default save behaviour without having to pass `--save`
2. You can override the config by adding `--save`
    - `--save` without a folder path defaults to `src/images/`
    - `--save ~/home/benny/Desktop/pattern` saves to specified location.
3. If folder does not exist, it will be created. 
4. Images are saved to folder by datetime (%d_%b_%y_%H%M) i.e `src/images/30_Jan_24_1215/`

Lastly there is a caveat. Currently the patterns will continue to be detected as long as the pattern is valid. If you run it daily, you would have an updated image of the same pattern. :smile_cat: 

I have an alternate implementation,

1. Store the pattern in a separate json file (state.json), when it is first detected.
2. On subsequent scans, check if the pattern exists in state.json file. If not, add it
3. If the pattern exists but is no longer detected, remove from pattern from state.json.
4. Now **only newly detected patterns** are displayed.
5. **To see all current patterns in market, run `py init.py --plot state.json`**

This will only apply if you pass the symbols as a watchlist file.

**Comment by BennyThadikaran** [comment]
Hi Saurabh, I have added this feature in the latest update. see #9

**Reply by cotton1234** [reply]
Thanks Benny. Let me check the same after eod file comes.

**Reply by cotton1234** [reply]
Benny, Seems i am not able to understand the same. Images directory got created but no image is generated. Am i missing the setup. 

saurabhgarg@MacBook-Pro src % ls -ltr
total 296
-rw-r--r--  1 saurabhgarg  staff      0 Jan 26 08:39 data.csv
-rw-r--r--  1 saurabhgarg  staff  22532 Jan 26 08:42 daily.csv
-rw-r--r--  1 saurabhgarg  staff  29434 Jan 26 17:42 custom_utils.py
-rw-r--r--  1 saurabhgarg  staff   5284 Jan 30 20:50 hnsu.json
-rw-r--r--  1 saurabhgarg  staff   6094 Jan 31 18:44 Plotter.py
-rw-r--r--  1 saurabhgarg  staff  12390 Jan 31 18:44 init.py
-rw-r--r--  1 saurabhgarg  staff  33495 Jan 31 18:44 utils.py
-rw-r--r--  1 saurabhgarg  staff  17895 Jan 31 18:57 bull.json
drwxr-xr-x  3 saurabhgarg  staff     96 Jan 31 19:08 images
-rw-r--r--  1 saurabhgarg  staff    211 Jan 31 19:12 user.json
drwxr-xr-x  3 saurabhgarg  staff     96 Jan 31 19:25 state
saurabhgarg@MacBook-Pro src % vi Plotter.py
saurabhgarg@MacBook-Pro src %
saurabhgarg@MacBook-Pro src % python3 init.py --plot bull.json

Chart Controls
        q: quit
        n: Next
        p: Previous
saurabhgarg@MacBook-Pro 31_Jan_24_1908 % pwd
/Users/saurabhgarg/shivshakti/stock-pattern/src/images/31_Jan_24_1908
saurabhgarg@MacBook-Pro 31_Jan_24_1908 % ls -ltr
total 0
saurabhgarg@MacBook-Pro 31_Jan_24_1908 %

**Comment by BennyThadikaran** [comment]
I got a fresh clone and ran the setup. Everything is working well. I think you ran the wrong commands.

1. Delete the state and image folders.
2. Run `py init.py --save`
3. Run the scan from the interactive menu.
4. If all went well, check the images folder.

`--plot` is used to load charts from the json files output during the scans. 

`py init.py -p all -f data.csv` will output `src/all.json` and `state/data_all.json`.

To plot the results again, `py init.py --plot all.json` or to check all existing patterns `py init.py --plot state/data_all.json`

If you dont want to use `--save` all the time, just add an entry `SAVE_FOLDER` to user.json with the folder path.

Hope that clarifies. Let me know

**Reply by cotton1234** [reply]
Works fine Benny !!
